flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSourceContexts.java

```diff
@@ -313,6 +325,7 @@ protected void processAndCollectWithTimestamp(T element, long timestamp) {
 		@Override
 		protected void processAndEmitWatermark(Watermark mark) {
 			output.emitWatermark(mark);
+			watermarkGauge.setCurrentLowWatermark(mark.getTimestamp());
```

Comment:
Should probably move this to the base abstract class `WatermarkContext`, after `processAndEmitWatermark(mark)` in the `emitWatermark` method (line):
https://github.com/apache/flink/pull/5092/files#diff-e38689efd0a97938646aa97eb8a97d96R438

The `processAndEmitWatermark` method of `AutomaticWatermarkContext` should actually also update the watermark metric after emitting the `Long.MAX_VALUE` watermark, but currently doesn't:
https://github.com/apache/flink/pull/5092/files#diff-e38689efd0a97938646aa97eb8a97d96R215

Implementing the metric update in the base class should cover both cases.

Reply:
I've thought about doing that, but when i noticed that it wouldn't work for the `AutomaticWatermarkContext` which emits watermarks in `processAndCollect`, and that contexts may emit watermarks anywhere (meaning we couldn't reliably cover all cases in the base class) i figured that we may just let each context do it separately.

Reply:
Yes, for the `AutomaticWatermarkContext` we still need some specific update logic in there for the periodic watermarks.

But we should still be able to migrate https://github.com/apache/flink/pull/5092/files/e2b20ee95ca66cbf42c7f3295863bc95605b4521#diff-e38689efd0a97938646aa97eb8a97d96R328 and the missing update after https://github.com/apache/flink/pull/5092/files#diff-e38689efd0a97938646aa97eb8a97d96R215 to the base class, no?